### PR TITLE
prepare for wal replayed successfully after restart

### PR DESCRIPTION
**Pull request by mychaow**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/service/IoTDB.java [hunk 1]
```diff
@@ -123,7 +120,22 @@ private void setUp() throws StartupException {
...
+
+    logger.info("IoTDB is set up, now may some sgs are not ready, don't worry, the superman Dr. Huang will teach the sg to fix themselves.");
+
+    while (!StorageEngine.getInstance().isAllSgReady()) {
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+    }
```
Comment (jt2594838) [inline server/src/main/java/org/apache/iotdb/db/service/IoTDB.java:132]: How about just waiting where you created that single thread pool and shuting down it?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 2]
```diff
@@ -792,8 +792,9 @@ private void tryToUpdateBatchInsertLastCache(InsertTabletPlan plan, Long latestF
           continue;
         }
         // Update cached last value with high priority
-        ((MeasurementMNode) manager.getChild(node, measurementList[i]))
-            .updateCachedLast(plan.composeLastTimeValuePair(i), true, latestFlushedTime);
+        Path tmpPath = new Path(plan.getDeviceId(), measurementList[i]);
+        manager.updateLastCache(tmpPath.getFullPath(),
+          plan.composeLastTimeValuePair(i), true, latestFlushedTime);
```
[the remark this answers is not in the document]
Comment (mychaow) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:797]: ok

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 3]
```diff
@@ -252,18 +267,24 @@ public StorageGroupProcessor getProcessor(String path) throws StorageEngineExcep
...
                 .getStorageGroupNode(storageGroupName);
-            processor.setDataTTL(storageGroup.getDataTTL());
-            processorMap.put(storageGroupName, processor);
+              processor.setDataTTL(storageGroup.getDataTTL());
+              processorMap.put(storageGroupName, processor);
+            }
           }
+        } else {
+          // not finished recover, refuse the request
+          throw new StorageEngineException("the sg " + storageGroupName + " may not ready now, please wait and retry later");
```
[the remark this answers is not in the document]
Comment (mychaow) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:287]: yes, some wait will be friendly, I will add it.

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 4]
```diff
@@ -135,6 +140,15 @@ private StorageEngine() {
     // recover upgrade process
     UpgradeUtils.recoverUpgrade();
 
+    recover();
+  }
+
+  public void recover() {
+    ExecutorService executors = Executors.newSingleThreadExecutor();
+    executors.submit(this::recoverAllSgs);
```
[the remark this answers is not in the document]
Comment (mychaow) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:148]: Wait util all sgs have been recoverd. This means all sgs could get the remote schema or someone return that the schema is not exist when replayed the wal

File: server/src/main/java/org/apache/iotdb/db/service/IoTDB.java (hunk 1 above)
Comment (mychaow) [reply server/src/main/java/org/apache/iotdb/db/service/IoTDB.java:132]: Good idea!

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 5]
```diff
@@ -1837,6 +1758,40 @@ public void cacheSchema(String path, MeasurementSchema schema) {
...
+
+  public TimeValuePair getLastCache(String seriesPath) {
+   try {
+     MeasurementMNode node = null;
+     node = (MeasurementMNode) mtree.getNodeByPath(seriesPath);
+     return node.getCachedLast();
+  } catch (MetadataException e) {
+     // do nothing
+  }
+  return null;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:1792]: please log the exception and format the code

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 6]
```diff
@@ -785,8 +785,12 @@ private void tryToUpdateBatchInsertLastCache(InsertTabletPlan plan, Long latestF
         continue;
       }
       // Update cached last value with high priority
-      ((MeasurementMNode) node.getChild(measurementList[i]))
-          .updateCachedLast(plan.composeLastTimeValuePair(i), true, latestFlushedTime);
+      MeasurementMNode tmpMeasurementNode = null;
+      if (node != null) {
+        tmpMeasurementNode = (MeasurementMNode) node.getChild(measurementList[i]);
+      }
+      IoTDB.metaManager.updateLastCache(node.getFullPath() + IoTDBConstant.PATH_SEPARATOR + measurementList[i],
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:792]: The String concat is an overhead in single node IoTDB, could we avoid this?

Comment (mychaow) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:792]: How about the following code?
` if (tmpMeasurementNode != null) {
        IoTDB.metaManager.updateLastCache(tmpMeasurementNode.getFullPath(),
          plan.composeLastTimeValuePair(i), true, latestFlushedTime, tmpMeasurementNode);
      } else {
        IoTDB.metaManager.updateLastCache(node.getFullPath() + IoTDBConstant.PATH_SEPARATOR + measurementList[i],
          plan.composeLastTimeValuePair(i), true, latestFlushedTime, tmpMeasurementNode);
      }`

**Pull request merged**